converter/interceptor.go

```diff
@@ -0,0 +1,215 @@
+		}
+	}
+
+	return nil
+}
+
+func (s *serviceInterceptor) processEvents(events []*historypb.HistoryEvent) error {
+	var err error
+
+	for _, e := range events {
+		switch e.EventType {
+		case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_STARTED:
+			err = decodePayloads(e.GetWorkflowExecutionStartedEventAttributes().Input, s.encoders...)
+		case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_COMPLETED:
+			err = decodePayloads(e.GetWorkflowExecutionCompletedEventAttributes().Result, s.encoders...)
+		case enumspb.EVENT_TYPE_START_CHILD_WORKFLOW_EXECUTION_INITIATED:
+			err = decodePayloads(e.GetStartChildWorkflowExecutionInitiatedEventAttributes().Input, s.encoders...)
+		case enumspb.EVENT_TYPE_SIGNAL_EXTERNAL_WORKFLOW_EXECUTION_INITIATED:
+			err = decodePayloads(e.GetSignalExternalWorkflowExecutionInitiatedEventAttributes().Input, s.encoders...)
+		case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_CONTINUED_AS_NEW:
+			err = decodePayloads(e.GetWorkflowExecutionContinuedAsNewEventAttributes().Input, s.encoders...)
+		case enumspb.EVENT_TYPE_ACTIVITY_TASK_SCHEDULED:
+			err = decodePayloads(e.GetActivityTaskScheduledEventAttributes().Input, s.encoders...)
+		case enumspb.EVENT_TYPE_ACTIVITY_TASK_COMPLETED:
+			err = decodePayloads(e.GetActivityTaskCompletedEventAttributes().Result, s.encoders...)
+		case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_SIGNALED:
+			err = decodePayloads(e.GetWorkflowExecutionSignaledEventAttributes().Input, s.encoders...)
+		case enumspb.EVENT_TYPE_CHILD_WORKFLOW_EXECUTION_COMPLETED:
+			err = decodePayloads(e.GetChildWorkflowExecutionCompletedEventAttributes().Result, s.encoders...)
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// NewPayloadEncoderGRPCServerInterceptor returns a GRPC Server Interceptor that will mimic the encoding
+// that the SDK system would perform when configured with a matching EncodingDataConverter.
+// Note: This approach does not support use cases that rely on the ContextAware DataConverter interface as
+// workflow context is not available at the GRPC level.
+func NewPayloadEncoderGRPCServerInterceptor(encoders ...PayloadEncoder) grpc.UnaryServerInterceptor {
+	s := serviceInterceptor{encoders: encoders}
+
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		err := s.processRequest(req)
+		if err != nil {
+			return nil, err
+		}
+
+		resp, err := handler(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+
+		return s.processResponse(resp), err
+	}
+}
+
+// NewPayloadEncoderGRPCClientInterceptor returns a GRPC Client Interceptor that will mimic the encoding
+// that the SDK system would perform when configured with a matching EncodingDataConverter.
+// Note: This approach does not support use cases that rely on the ContextAware DataConverter interface as
+// workflow context is not available at the GRPC level.
+func NewPayloadEncoderGRPCClientInterceptor(encoders ...PayloadEncoder) grpc.UnaryClientInterceptor {
```

Comment:
Might want to take an options struct here like you did w/ the proxy for future proofing (and return an error).